Added NonlinearSusceptibility with chi3 support

caseyflex opened the pull request:
Known limitations:

- Currently only supports non-dispersive Medium

tylerflex commented on tidy3d/components/source.py:
```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
 
-SourceTimeType = Union[GaussianPulse, ContinuousWave]
+class CustomSourceTime(Pulse):
+    """Custom source time dependence.
```
I think this and the note require a bit more discussion. So if a user makes a CustomTimeSource with some arbitrary values `v_i` placed at times `t_i`, the source that the simulation sees at time `t` will just be the `v_i` corresponding to the nearest time coordinate? Wouldn't this have the potential to cause a source time dependence that isn't smooth (and therefore contains very high frequency components?). If so, is there a way to mitigate this? (eg by providing a `dt` that is much smaller than the simulation `dt`?  I just think we need to make it explicit what this object is doing on the solver side, eg how the source value is chosen at time `t` given the data points, and also perhaps discuss these edge cases (like where the data points are not the same as the simulation time) and potential issues.

caseyflex replied:
yeah, using a small `dt` migitages this. Agreed, need more discussion here. I'll also split off the PR for CustomSourceTime

tylerflex commented on tidy3d/components/source.py:
```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
+    >>> sourcetime = CustomSourceTime.from_numpy(values=np.linspace(0, 1, 100), dt=.1)
+    """
+
+    sourcetime_dataset: Optional[TimeDataset] = pydantic.Field(
```
the `Source` has `source_time` as a field, so I'd vote we change this to `source_time_dataset` if it doesn't break any PEP8 convention for variable naming.

tylerflex commented on tidy3d/components/source.py:
```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
+
+    Example
+    -------
+    >>> sourcetime = CustomSourceTime.from_numpy(values=np.linspace(0, 1, 100), dt=.1)
```
does this work? what about freq0 and fwidth?

tylerflex commented on tidy3d/components/source.py:
```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
+    def from_numpy(
+        cls, freq0: float, fwidth: float, values: ArrayComplex1D, dt: float
+    ) -> "CustomSourceTime":
+        """Create a CustomSourceTime from a numpy array.
```
I think it needs some more explicit  mention of what time coordinate are used (assume the user isn't looking at the source) for example.

```
The time coordinates used are evenly spaced between 0 and dt * (N-1) with a step size of `dt` where N is the length of the values array.

tylerflex replied:
maybee line 342 is a good place for this.

tylerflex commented on tidy3d/components/source.py:
```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
+    _sourcetime_dataset_none_warning = warn_if_dataset_none("sourcetime_dataset")
+
+    @classmethod
+    def from_numpy(
```
maybe `from_values` or something makes more sense here? One could imagine `from_numpy` perhaps accepting the time coordinates instead of dt, for example.

tylerflex commented on tidy3d/components/source.py:
```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
+    @classmethod
+    def from_numpy(
+        cls, freq0: float, fwidth: float, values: ArrayComplex1D, dt: float
+    ) -> "CustomSourceTime":
```
no need to put this in quotes if there's an 
```py
from __future__ import annotations
```
line at the top of the file

tylerflex commented on tidy3d/components/source.py:
```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
+        Returns
+        -------
+        CustomSourceTime
+            `CustomSourceTime` with these values.
```
please use :class: linking for the CustomSourceTime

tylerflex commented on tidy3d/components/source.py:
```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
+        return CustomSourceTime(
+            freq0=freq0,
+            fwidth=fwidth,
+            sourcetime_dataset=TimeDataset(values=TimeDataArray(values, coords=dict(t=times))),
```
can we construct this variable before passing it to `CustomSourceTime` for readability?

```py
sourcetime_dataset=TimeDataArray(values, coords=dict(t=times))

return CustomSourceTime(
         freq0=freq0,
         fwidth=fwidth,
         sourcetime_dataset=sourcetime_dataset,
)

tylerflex commented on tidy3d/components/source.py:
```diff
@@ -9,10 +9,20 @@
+from .types import (
+    Direction,
+    Polarization,
+    Ax,
```
please split this import into two lines
```py
from .types import x, y, z
from .types import a, b, c
```

just an arbitrary code style thing, for consistency.

tylerflex commented on tidy3d/components/source.py:
```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
+        ..., title="Source time dataset", description="Dataset for storing the custom source time."
+    )
+
+    _sourcetime_dataset_none_warning = warn_if_dataset_none("sourcetime_dataset")
```
do we want to add any validation on the supplied sourcetime_dataset? For example, what if the time coordinates are all negative? or all extend outside of the simulation run_time? will it be handled as expected?

tylerflex commented on tests/test_components/test_IO.py:
```diff
@@ -37,6 +37,8 @@ def set_datasets_to_none(sim):
             src["field_dataset"] = None
         elif src["type"] == "CustomCurrentSource":
             src["current_dataset"] = None
+        if src["source_time"]["type"] == "CustomSourceTime":
```
I think we also need to add this to the `Simulation.custom_datasets` property for things to work properly. For example, if not added there, the simulation may not be uploaded properly as an hdf5

https://github.com/flexcompute/tidy3d/blob/develop/tidy3d/components/simulation.py#L2654-L2670

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -2927,6 +2927,51 @@ class CustomAnisotropicMediumInternal(CustomAnisotropicMedium):
+class AbstractNonlinearMedium(AbstractMedium, ABC):
+    """Abstract nonlinear medium.
+
+    Note
```
not sure this note will propagate to the subclasses, so might need to copy it over.

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -2927,6 +2927,51 @@ class CustomAnisotropicMediumInternal(CustomAnisotropicMedium):
+    Note
+    ----
+    For complex fields (e.g. when using Bloch boundary conditions), the nonlinearity
+    is applied separately to the real and imaginary parts.
```
could you be more explicit? so does it mean that the P_NL imaginary and real parts are computed separately from |real(E)|^3 and  |imag(E)|^3 respectively? or something else? what are the practical considerations of this? for example, is it still accurate to use KerrMedium with bloch boundaries?

caseyflex replied:
yeah, it's still accurate to use KerrMedium with bloch boundaries. It wouldn't be physical do use the complex fields directly in definining the instantaneous nonlinearity, since the instantaneous physical fields are real.

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -2927,6 +2927,51 @@ class CustomAnisotropicMediumInternal(CustomAnisotropicMedium):
+    Note
+    ----
+    Different field components do not interact nonlinearly. For example,
+    when calculating :math:`P_{NL}_x`, we approximate :math:`|E|^2 \\approx |E_x|^2`.
```
this seems potentially quite limiting. does this mean that the nonlinearity is only accurate for 2D simulations with E field out of plane? (for example?) again, what are the practical ramifications.

caseyflex replied:
It's accurate as long as the E field is predominantly polarized along one of the Cartesian axes. See also the discussion at the backend PR.

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -2927,6 +2927,51 @@ class CustomAnisotropicMediumInternal(CustomAnisotropicMedium):
+    >>> medium = KerrMedium(chi3=1)
+    """
+
+    chi3: float = pd.Field(..., title="Chi3", description="Chi3 nonlinear susceptibility.")
```
can chi3 be negative? (I forget...) if not, make this a `pd.NonNegativeFloat` type.

caseyflex replied:
yeah, I think we should allow it to be negative

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -2927,6 +2927,51 @@ class CustomAnisotropicMediumInternal(CustomAnisotropicMedium):
+    """
+
+    numiters: pd.PositiveInt = pd.Field(
+        1,
```
is 1 a reasonable number or should this be adjusted case by case? it might be good to start with a default value of something more practical for most cases (define as a constant at the top of file)? but maybe I'm not understanding something.

caseyflex replied:
1 has been working fine for me. I think it might need to be adjusted if the nonlinearity was quite strong.

tylerflex reviewed (changes requested):
Thanks @caseyflex, I think both of these will be super exciting to have. I suggest having @momchil-flex chime in specifically on how we do interpolation of the custom source time as I'm a bit concerned about edge cases using nearest interpolation.  Also, is it possible to split this into two PRs? they seem quite separate no? or at least two commits if possible. Otherwise it will be a bit confusing to have both somewhat orthogonal changes in a single commit.

tylerflex commented:
Regarding this:

> Can we make this subclass something that could be either Medium or DispersiveMedium?

I dont see why not. currently the Kerr Medium assumes linear medium properties of vacuum? or is this a WIP?

caseyflex commented on tidy3d/components/source.py:
```diff
@@ -291,7 +301,72 @@ def amp_time(self, time: float) -> complex:
+        """
+        if self.sourcetime_dataset is None:
+            return None
+        return self.sourcetime_dataset.values.sel(t=time, method="nearest").to_numpy()
```
When I was using this, I was oversampling enough in my input signal that the interpolation method didn't matter. We could change it though, if you think e.g. linear interpolation is less likely to cause problems.

caseyflex commented:
CustomSourceTime moved here: https://github.com/flexcompute/tidy3d/pull/994

tylerflex commented:
Hm, maybe we can stick with `Medium` as the superclass until someone requests dispersive nonlinear? Otherwise we could I guess consider two `KerrMedium` and `KerrPoleResidue` which both inherit from the abstract nonlinear medium and then `Medium`, `PoleResidue` respectively? For now probably just `Medium` is fine.

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -2927,6 +2927,61 @@ class CustomAnisotropicMediumInternal(CustomAnisotropicMedium):
+    For complex fields (e.g. when using Bloch boundary conditions), the nonlinearity
+    is applied separately to the real and imaginary parts, so that the above equation
+    holds when both E and :math:`P_{NL}` are replaced by their real or imaginary parts.
+    Since the physical fields are real, it wouldn't make sense to apply the instantaneous
```
maybe instead of this last sentence with "it wouldn't make sense" we can soften it with something like "The nonlinearity is only applied to the real-valued fields since they are the physical fields."

tylerflex reviewed (approved):
This looks good to me pending two minor rewordings, thanks @caseyflex!

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -2927,6 +2927,61 @@ class CustomAnisotropicMediumInternal(CustomAnisotropicMedium):
+    Note
+    ----
+    The nonlinear constitutive relation is solved iteratively; it may not converge
+    for large nonlinearities. Increasing `numiters` can help with convergence.
```
perhaps "strong" is better than "large"?

momchil-flex reviewed (commented):
Looks good!

> Backend supports dispersive KerrMedium but frontend currently does not. Can we make this subclass something that could be either Medium or DispersiveMedium?

This goes back to our discussion on slack today... It really does feel like our medium subclassing is becoming quite a mess. One option is to treat the nonlinear susceptibility (and the time-modulated permittivity in the future @weiliangjin2021 ) as something that can be added to any medium where it's supported, e.g.

```
td.Medium(permittivity=..., nonlinearity=KerrNonlinearity(chi3=1))
```

Then we can pick which media have a `nonlinearity` Field. In fact this could still be done through subclassing, something like

```
# this would probably still subclass from some abstract nonlinearity class
class KerrNonlinearity():
    chi3: float
    numiters: pd.PositiveInt

class AbstractNonlinearMedium(AbstractMedium, ABC):
    nonlinearity: Union[KerrNonlinearity, ...]

# now AbstractNonlinearMedium can be added anywhere we support, e.g.
class Medium(AbstractMedium, AbstractNonlinearMedium):
    ...

class DispersiveMedium(AbstractMedium, AbstractNonlinearMedium, ABC):
    ...

```

In the case of nonlinearity, can it pretty much be added on top of any medium, including custom?

momchil-flex commented:
Note: this is just a suggestion, not saying you should implement this right now. But we should decide on something (@tylerflex )

momchil-flex commented:
Whatever way we introduce the nonlinear medium, we should make sure the mode solver issues a warning that the nonlinearity is not used if someone tries a mode solve in a plane that contains a nonlinear medium. I forget, how do we handle 2D media in the mode solver now?

tylerflex commented:
> This goes back to our discussion on slack today... It really does feel like our medium subclassing is becoming quite a mess. One option is to treat the nonlinear susceptibility (and the time-modulated permittivity in the future @weiliangjin2021 ) as something that can be added to any medium where it's supported
> But we should decide on something (@tylerflex )

In principle it makes sense to be able to add nonlinearity to any medium. However, with the way things are set up with subclassing, I think it might make things a bit complicated. My thinking is just make KerrMedium [and eventually KerrDispersiveMedium], let the class type do the work regarding what properties the medium has, but eventually we should just separate all of the **categories** of effects in the medium:
* Linear properties
* Nonlinear properties
* Dispersion
* Time dependence
into their own fields of a Medium class, each determined by their own specs, and handle custom medium by overloading any fields with data arrays. but this seems like a 3.0 change.

I think how Casey has it here is good for now and we can always add the nonlinearity fields into the AbstracMedium class without breaking backwards compatibility if we change our minds?

caseyflex commented:
> Whatever way we introduce the nonlinear medium, we should make sure the mode solver issues a warning that the nonlinearity is not used if someone tries a mode solve in a plane that contains a nonlinear medium. I forget, how do we handle 2D media in the mode solver now?

We use `volumetric_structures` in `Simulation.epsilon_on_grid` and related functions, so the mode solver should work correctly for 2D materials. To handle nonlinear case, I'll put a warning in `Simulation.epsilon_on_grid`. Of course there's no way to know epsilon a priori for nonlinear materials.

caseyflex commented:
> > This goes back to our discussion on slack today... It really does feel like our medium subclassing is becoming quite a mess. One option is to treat the nonlinear susceptibility (and the time-modulated permittivity in the future @weiliangjin2021 ) as something that can be added to any medium where it's supported
> > But we should decide on something (@tylerflex )
> 
> In principle it makes sense to be able to add nonlinearity to any medium. However, with the way things are set up with subclassing, I think it might make things a bit complicated. My thinking is just make KerrMedium [and eventually KerrDispersiveMedium], let the class type do the work regarding what properties the medium has, but eventually we should just separate all of the **categories** of effects in the medium:
> 
> * Linear properties
> * Nonlinear properties
> * Dispersion
> * Time dependence
> 
> into their own fields of a Medium class, each determined by their own specs, and handle custom medium by overloading any fields with data arrays. but this seems like a 3.0 change.
> 
> I think how Casey has it here is good for now and we can always add the nonlinearity fields into the AbstracMedium class without breaking backwards compatibility if we change our minds?

Yeah, let's save the restructuring for 3.0. I'll leave it this way with `KerrMedium` for now, and introduce other classes like `KerrDispersiveMedium` as needed later.

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -71,6 +71,53 @@ def _eps_model(self, frequency: float) -> complex:
 
 
+class Chi3Spec(Tidy3dBaseModel):
+    """Chi3 medium described by a chi3 nonlinear susceptibility.
```
"Specification adding chi3 nonlinear susceptibility to a medium." probably is more accurate.

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -71,6 +71,53 @@ def _eps_model(self, frequency: float) -> complex:
 """ Medium Definitions """
 
 
+class Chi3Spec(Tidy3dBaseModel):
```
Do we want an abstract `NonlinearSpec`? or I guess maybe later?

caseyflex replied:
maybe later I think

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -93,8 +140,23 @@ class AbstractMedium(ABC, Tidy3dBaseModel):
+        """Validate compatibility with the nonlinear spec."""
+        if val is None:
+            return None
+        if issubclass(cls, Medium):
```
I think this is a problematic way to do this. It's a bit circular? if we moved `Medium` and `AbstractMedium` to their own files, it would break I think. We're dealing with some issues like this in `geometry.py` right now and it's a headache so I'd say let's keep AbstratMedium agnostic to Medium.

A few possible solutions:

1. only move `nonlinear_spec` to the `Medium` for now until we find a better way to do this.

2.  add `nonlinear_spec` to `AbstractMedium` like this, but only allow `None` values.  if it's not None, give the warning that you have here (optional).  Add `nonlinear_spec`  to `Medium` with a allowed type of `Union[Chi3Spec]` and overwrite validator if we had one.

In general, we need a simple way to validate whether we support different possible combinations of specs.  Haven't figured out a good solution yet.

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -3186,6 +3244,10 @@ def perturbed_copy(
         return CustomPoleResidue.parse_obj(new_dict)
 
 
+# PEC builtin instance
```
curious why did this get moved?

tylerflex commented on tidy3d/components/simulation.py:
```diff
@@ -2885,6 +2885,12 @@ def make_eps_data(coords: Coords):
                     red_coords = Coords(**dict(zip("xyz", coords_reduced)))
                     eps_structure = get_eps(structure=structure, frequency=freq, coords=red_coords)
 
+                    if not structure.medium.nonlinear_spec is None:
```
```py
if not structure.medium.nonlinear_spec is None:
```
becomes
```py
if structure.medium.nonlinear_spec is not None:
```
is the convention in python.

tylerflex commented:
This will be a better way to do things, I can already tell. Thanks @caseyflex !

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -71,6 +71,50 @@ def _eps_model(self, frequency: float) -> complex:
+    >>> medium = Medium(permittivity=2, nonlinear_spec=Chi3Spec(chi3=1))
+    """
+
+    chi3: float = pd.Field(..., title="Chi3", description="Chi3 nonlinear susceptibility.")
```
does this have units?

dbochkov-flexcompute commented on tidy3d/components/medium.py:
```diff
@@ -93,6 +137,12 @@ class AbstractMedium(ABC, Tidy3dBaseModel):
         "useful in some cases.",
     )
 
+    nonlinear_spec: None = pd.Field(
```
just wanted to clarify how this works: would the user just get an error if they try to plug `Chi3Spec` into a dispersive medium? Ideally we should have some meaningful error message something like `f"Nonlinear feature is not currently supported for medium class {type(cls)}"`, but it's up to you

Btw, `CustomIsotropicMedium` (which will become just `CustomMedium` in 3.0) inherits from `Medium`. Do we support `Chi3Spec` for custom mediums on backend?

dbochkov-flexcompute reviewed (approved):
Looks good to me too!

caseyflex commented:
I added AbstractNonlinearSpec for convenience later, and I also changed AbstractMedium to take NonlinearSpecType, but then validate it as needed for different medium types. I think this addresses Daniil's concern and is more robust for the future.

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -71,6 +74,75 @@ def _eps_model(self, frequency: float) -> complex:
+        """Check that numiters is not too large."""
+        if val > NONLINEAR_MAX_NUMITERS:
+            raise ValidationError(
+                "NonlinearSpec.numiters must be less than "
```
let's put anything code related in single quotes `'`. `'NonlinearSpec.numiters'`.

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -93,6 +165,22 @@ class AbstractMedium(ABC, Tidy3dBaseModel):
+        description="Nonlinear spec applied on top of the base medium properties.",
+    )
+
+    @pd.validator("nonlinear_spec", always=True)
```
feel like we need a better way to do this more generally. so let's say we have 
```py
NonlinearSpecType = Union[None, Spec1, Spec2, ..., SpecN]
```
and then classes `ClassA`, `ClassB`, ... that each can only accept a certain subset of `NonlinearSpecType`.

Maybe the most general way is to just define in the `AbstractMedium`
```py
nonlinear_spec : None = pd.Field(None, ...)
```
and then in the individual classes
```py
nonlinear_spec: Union[None, Spec1]
```
? Which maybe becomes a bit annoying as the combinations grow but is reasonably general? Would pydantic take care of erroring if one tried to put the wrong spec in?  Basically I feel it might be nice to do this only in the type annotation rather than the validator. This also makes it more clear in the docs.

tylerflex replied:
On the other hand, if everything is merged into the same medium, it might be harder to keep track of the dependencies just in the type annotation. Say a non-dispersive medium can accept one type of NonlinearSpec but a dispersive can only accept another. There would be some cross coupling between the two specs that I think could only be handled through validators, so not sure. Any thoughts?

caseyflex replied:
My opinion is that the type annotations are not the best tool for this. The incompatibilities seem to usually be from things that haven't been implemented yet, and cross coupling between different specs (for example, once pole residue is a spec, nonlinear + dispersive)

caseyflex replied:
If we did go with the type annotation approach, like Daniil mentioned, we would still want validators that would give more helpful error messages in certain cases, like when Chi3Spec is added to CustomIsotropicMedium. So my feeling is that the type annotations add boilerplate without much content

caseyflex replied:
but the real argument is actually like you said, the cross coupling between different specs down the line.

tylerflex replied:
true. If we had a set of "allowed" combinations, though. we could always have a generic validator like
```
def validate_specs(cls, values):
    nonlinear_spec_type = type(values.get("nonlinear_spec"))
    dispersion_spec_type = type(values.get("dispersion_spec"))
    if (nonlinear_spec_type, dispersion_spec_type) not in ALLOWED_SPEC_COMBINATIONS:
         raise ValidationError(f"Combination of {nonlinear_spec_type} and {dispersion_spec_type} not currently supported.").
```
which could be cleaner and we could document the `ALLOWED_SPEC_COMBINATIONS`. What I worry about slightly is a set of validators with really convoluted logic and no simple way for the user to see what combinations are actually allowed. However, I guess if the allowed combinations is pretty easy to write down as a function, it might indeed be easier? @dbochkov-flexcompute any thoughts?

tylerflex commented on tidy3d/components/medium.py:
```diff
@@ -71,6 +74,75 @@ def _eps_model(self, frequency: float) -> complex:
+    """
+
+    chi3: float = pd.Field(
+        ..., title="Chi3", description="Chi3 nonlinear susceptibility, in units of um^2 / V^2."
```
we should pass the units as a keyword argument to `Field`, eg.

```py
chi3: float = pd.Field(
    ...,
    title=...,
    description=...,
    units="um^2 / V^2",
)
```
We parse this field when setting units in the API reference. Could you add this? Thanks.

tylerflex reviewed (approved):
Approve pending the units kwarg in `pd.Field()`. thanks @caseyflex
